presto-matching/src/main/java/com/facebook/presto/matching/Property.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.matching;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+public interface Property<F, T>
+{
+    <R> PropertyPattern<F, R> matching(Pattern<R> pattern);
+
+    static <F, T> Property<F, T> property(Function<F, T> property)
+    {
+        return optionalProperty(source -> Optional.of(property.apply(source)));
+    }
+
+    static <F, T> Property<F, T> optionalProperty(Function<F, Optional<T>> property)
+    {
+        return new Property<F, T>()
+        {
+            @Override
+            public <R> PropertyPattern<F, R> matching(Pattern<R> pattern)
+            {
+                return PropertyPattern.of(property, pattern);
+            }
+        };
+    }
+
+    default PropertyPattern<F, T> capturedAs(Capture<T> capture)
```

Comment:
I did `property`, `optionalProperty`, `matching`, `capturedAs` to reflect the factories - abstract instance methods - instance methods order